File: nodejs/src/lcd.ts
import {EOL} from 'os';

const LCDNumbers = {
  "0": { first: " _ ", second: "| |", third: "|_|" },
  '1': { first: '   ', second: '  |', third: '  |'},
  '2': { first: ' _ ', second: ' _|', third: '|_ '},
  '3': { first: ' _ ', second: ' _|', third: ' _|'},
  '4': { first: '   ', second: '|_|', third: '  |'},
  '5': { first: ' _ ', second: '|_ ', third: ' _|'},
  '6': { first: ' _ ', second: '|_ ', third: '|_|'},
  '7': { first: ' _ ', second: '  |', third: '  |'},
  '8': { first: ' _ ', second: '|_|', third: '|_|'},
  '9': { first: ' _ ', second: '|_|', third: '  |'}
};

/**
 * Removes non numeric characters from characters list.
 *
 * @param charList - The initial characters list
 * @returns the charlist as string array purged from non numeric characters
 */
function sanitize(charList: string): Array<string> {
  return charList
    .split('')
    .filter(character => LCDNumbers[character]);
}

/**
 * Extracts a line of the LCD format from a character list.
 *
 * @param charList - The input characters list
 * @param property - the property to extract from character list
 * @returns a single line built from the given character list for the
 * property given into parameter.
 */
function extractLine(charList: Array<string>, property: string): string {
  return charList
    .map(character => LCDNumbers[character][property])
    .join(' ');
}

/**
 * Builds a LCD formated string
 *
 * @param input - The classic string to format to LCD display
 * @returns A LCD formated string
 */
export function buildLCDString(input: string = ''): string {
  const sanitizedList = sanitize(input);

  // Half functionnal style :)
  // Actually, with currification we could have chained the call of sanitize
  // with other function calls, but function currification is ugly in JS
  // and produces uncommon code. So let's not do it !
  return EOL + ["first", "second", "third"]
    .map(prop => extractLine(sanitizedList, prop))
    .join(EOL);
}

export default buildLCDString;

File: elixir/lib/lcd.ex
defmodule LCD do
  @moduledoc """
  Documentation for Lcd.
  """

  @lcd_numbers %{
    "0" => {" _ ", "| |", "|_|"},
    "1" => {"   ", "  |", "  |"},
    "2" => {" _ ", " _|", "|_ "},
    "3" => {" _ ", " _|", " _|"},
    "4" => {"   ", "|_|", "  |"},
    "5" => {" _ ", "|_ ", " _|"},
    "6" => {" _ ", "|_ ", "|_|"},
    "7" => {" _ ", "  |", "  |"},
    "8" => {" _ ", "|_|", "|_|"},
    "9" => {" _ ", "|_|", "  |"}
  }

  defp sanitize(charlist) do
    String.split(charlist, "")
    |> Enum.filter(fn c -> @lcd_numbers[c] != nil end)
  end

  defp extract(line, charlist) do
    charlist
    |> Enum.map(fn c -> @lcd_numbers[c] end)
    |> Enum.map(fn list -> elem(list, line) end)
    |> Enum.join("")
  end

  def to_lines(charlist) do
    0..2
    |> Enum.reduce("", fn i, acc -> acc <> extract(i, charlist) <> "\n" end)
  end

  @doc """
  Tranforms any string in LCD formated string.

  ## Examples
      iex> LCD.transform("12345")
      "    _  _     _ 
        | _| _||_||_ 
        ||_  _|  | _|
      "

      iex> LCD.transform("1a23*45")
      "    _  _     _ 
        | _| _||_||_ 
        ||_  _|  | _|
      "

  """
  def transform(inputString) do
    inputString
    |> sanitize
    |> to_lines
  end
end

defmodule LCD.CLI do
  def main(args) do
     LCD.transform(hd(args))
     |> IO.puts
  end
end
